File: backend/app/schemas/integration.py
"""Схемы раздела «Интеграция ФГИС» (только супер-админ).

Формы ответов строго по контракту эндпоинтов /api/v1/integration.
"""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class RegionsOverview(BaseModel):
    """Сводка по справочнику регионов."""
    total: int
    last_sync: Optional[datetime] = None  # max(Region.last_sync)


class MnoOverview(BaseModel):
    """Сводка по МНО."""
    total: int


class PerRegionStat(BaseModel):
    """Строка сводной таблицы: сколько МНО в базе по региону + дата синхронизации."""
    code: str
    name: str
    fed: int
    mno_count: int
    last_sync: Optional[datetime] = None


class IntegrationOverview(BaseModel):
    """GET /integration/overview."""
    regions: RegionsOverview
    mno: MnoOverview
    per_region: list[PerRegionStat]


class RegionsSyncResult(BaseModel):
    """POST /integration/regions/sync — итог синхронизации справочника регионов."""
    total: int
    created: int
    updated: int
    last_sync: datetime


class MnoSyncRequest(BaseModel):
    """Тело POST /integration/mno/sync."""
    region_code: str = Field(min_length=1)


class MnoSyncStartResult(BaseModel):
    """Ответ POST /integration/mno/sync — принятая (или уже идущая) фоновая задача."""
    job_id: str
    region_code: str
    state: str  # running


class MnoSyncStatus(BaseModel):
    """GET /integration/mno/sync/status — прогресс/итог фоновой синхронизации МНО."""
    job_id: str
    region_code: str
    region_name: str
    state: str  # running | done | error
    discovered: int
    fetched: int
    upserted: int
    error: Optional[str] = None
    started_at: datetime
    finished_at: Optional[datetime] = None
    # Прогон по ВСЕМ регионам (scope="all"): порегионный прогресс. Для одиночной
    # задачи дефолты держат прежнюю форму (scope="region", один регион).
    scope: str = "region"          # "region" | "all"
    regions_total: int = 1
    regions_done: int = 0
    regions_failed: int = 0
    current_region: str = ""

File: frontend/src/api/hooks/integration.ts
import { useQuery } from '@tanstack/react-query';
import { api } from '@/api/client';
import type { IntegrationOverview, MnoSyncStatus } from '@/api/aliases';

/**
 * Хуки чтения раздела «Интеграция ФГИС» (только супер-админ).
 * Сводка и статус фоновой синхронизации МНО — серверное состояние через TanStack Query.
 */

/**
 * GET /integration/overview — сводка: итоги по регионам/МНО + построчная сводка по регионам.
 * Инвалидируется после синхронизации регионов и по завершении фоновой синхронизации МНО.
 */
export function useIntegrationOverview() {
  return useQuery({
    queryKey: ['integration', 'overview'],
    queryFn: async () => {
      const res = await api.get<IntegrationOverview>('/integration/overview');
      return res.data;
    },
  });
}

/**
 * GET /integration/mno/sync/status?job_id — статус фоновой задачи синхронизации МНО.
 * Опрос ~1500мс, ПОКА state==='running' (опрос гасится сам на done/error); при null jobId
 * запрос выключен. Прогресс (discovered/fetched/upserted) держит Query, не локальный state.
 */
export function useMnoSyncStatus(jobId: string | null) {
  return useQuery({
    queryKey: ['integration', 'mno-sync', jobId],
    queryFn: async () => {
      const res = await api.get<MnoSyncStatus>(
        `/integration/mno/sync/status?job_id=${encodeURIComponent(jobId ?? '')}`
      );
      return res.data;
    },
    enabled: jobId != null,
    // Пока задача выполняется — перезапрашиваем; на done/error интервал выключается.
    refetchInterval: (query) => (query.state.data?.state === 'running' ? 1500 : false),
  });
}

/**
 * GET /integration/mno/sync/active — ЛЮБАЯ идущая фоновая синхронизация (регион ИЛИ «все»).
 * Нужен для переподключения к ней после перезагрузки страницы (F5): бэк-задача серверная и
 * НЕ прерывается, теряется лишь job_id во фронт-состоянии. Раньше подхватывался только прогон
 * «Все регионы», а синк одного региона терялся → UI предлагал запуск заново. Нет активной
 * задачи → бэк отвечает null (это НЕ ошибка).
 */
export function useRunningAnyJob() {
  return useQuery({
    queryKey: ['integration', 'mno', 'running-any'],
    queryFn: async (): Promise<MnoSyncStatus | null> => {
      const res = await api.get<MnoSyncStatus | null>('/integration/mno/sync/active');
      return res.data ?? null;
    },
    staleTime: 0,
  });
}
